lib/Parse/ParseExpr.cpp

```diff
@@ -2612,18 +2615,25 @@ ParserStatus Parser::parseClosureSignatureIfPresent(
 
       // Parse the func-signature-result, if present.
       if (consumeIf(tok::arrow)) {
+        sawTopLevelArrowInLookahead = true;
         if (!canParseType())
           return makeParserSuccess();
-
+        
         consumeEffectsSpecifiers();
       }
     }
 
     // Parse the 'in' at the end.
-    if (Tok.isNot(tok::kw_in))
-      return makeParserSuccess();
+    if (Tok.isNot(tok::kw_in)) {
+      // Even if 'in' is missing, the presence of '->' makes this look
+      // like a closure signature. There's no other valid syntax that
+      // could legally contain '->' at this position.
+      bool shouldTreatAsClosure = sawTopLevelArrowInLookahead;
```

Comment:
I don't think `shouldTreatAsClosure` is needed here. Let's just use `sawTopLevelArrowInLookahead`.